templates/rustdoc.hbs

```diff
@@ -16,19 +16,33 @@
 <script type="text/javascript" src="/menu.js?{{cratesfyi_version_safe}}"></script>
 </body>
 <script>
-    var doc_body = document.getElementById("rustdoc_body_wrapper");
-    if(window.location.hash) {
-        var notFirefox = typeof InstallTrigger === 'undefined';
-        if(notFirefox) {
-            var hash = window.location.hash;
-            window.location.hash = "";
-            setTimeout(function () {
-                window.location.hash = hash;
-                doc_body.focus();
-            }, 1);
-        }
-    } else {
-        doc_body.focus();
+  // Reset the scroll offset on browsers that don't support
+  // scroll-padding-top (Desktop & Mobile Safari):
+  const maybeFixupViewPortPosition = function() {
+    if (window.location.hash) {
+      const anchorElement = document.getElementById(window.location.hash.substr(1));
+      const navBarHeight = document.getElementsByClassName("nav-container-rustdoc")[0].offsetHeight;
+      if (anchorElement &&
+          anchorElement.getBoundingClientRect().top <= navBarHeight &&
+          Math.abs(anchorElement.getBoundingClientRect().top) >= 0) {
+        // It's just overlapped by the nav bar. This can't be a coincidence, scroll it into view:
+        requestAnimationFrame(function() {
+          scrollBy(0, -navBarHeight);
+        });
+      }
     }
+  };
+  window.addEventListener("hashchange", maybeFixupViewPortPosition, false);
+  // Fix up the scroll position if this was a direct visit to the page
+  // (not back/forward navigation):
+  if(window.performance) {
```

Comment:
What does looking at `window.performance` do?

nit:
```suggestion
  if (window.performance) {
```

Reply:
> What does looking at `window.performance` do?

This is for detecting whether the page was visited via the back/forward buttons: we should avoid scrolling to the anchor in that case, as people will expect to be returned to the place where they left the page in history.